borg/archive.py

```diff
@@ -467,23 +469,53 @@ def rename(self, name):
         self.cache.chunk_decref(self.id, self.stats)
         del self.manifest.archives[self.name]
 
-    def delete(self, stats, progress=False):
-        unpacker = msgpack.Unpacker(use_list=False)
-        items_ids = self.metadata[b'items']
-        pi = ProgressIndicatorPercent(total=len(items_ids), msg="Decrementing references %3.0f%%", same_line=True)
-        for (i, (items_id, data)) in enumerate(zip(items_ids, self.repository.get_many(items_ids))):
+    def delete(self, stats, progress=False, forced=False):
+        class FatalError(Exception):
+            """some fatal error occurred, the transaction must be aborted."""
+
+        def chunk_decref(id, stats):
+            nonlocal error
+            try:
+                self.cache.chunk_decref(id, stats)
+            except KeyError as e:
+                # chunkid missing from chunks index - chunks cache corrupted!
+                raise FatalError(str(e))
+            except Repository.ObjectNotFound as e:
+                # object not in repo - strange, but we wanted to delete it anyway.
+                # XXX doesn't this also mean that our chunks cache is corrupted?
+                # XXX obviously it had a refcount of 1 before decref, but the chunk
+                # XXX has already been deleted from the repo.
+                error = True
+
+        error = False
+        try:
+            unpacker = msgpack.Unpacker(use_list=False)
+            items_ids = self.metadata[b'items']
+            pi = ProgressIndicatorPercent(total=len(items_ids), msg="Decrementing references %3.0f%%", same_line=True)
+            for (i, (items_id, data)) in enumerate(zip(items_ids, self.repository.get_many(items_ids))):
+                if progress:
+                    pi.show(i)
+                unpacker.feed(self.key.decrypt(items_id, data))
+                chunk_decref(items_id, stats)
+                for item in unpacker:
+                    if b'chunks' in item:
+                        for chunk_id, size, csize in item[b'chunks']:
```

Comment:
What *may* be possible is that we don't get the expected types. That would be the long-metadata case, say a long list begins that would span multiple chunks but then one of these are dropped, msgpack tries hard to decode and usually succeds. Then the list elements could be totally different from what'd we expect, for example another item dict.